gravitational/teleport
# Connect: Implement SetGatewayLocalPort RPC

We want the user to be able to change the randomly assigned port number of a gateway (gravitational/webapps.e#302).

To accomplish this, we need an RPC that's going to change the port number and restart the gateway under a new port. That should happen only if the new port is actually free – if not then the RPC should fail and the gateway should continue running under the old port.

The PR might seem bigger than it actually is due to protobuf changes which are encapsulated within a single commit. Best reviewed commit-by-commit.

---

* Why not make an RPC called `UpdateGateway` instead of having both `SetGatewayTargetSubresourceName` and `SetGatewayLocalPort`?
   * Initially we had a slightly different idea on how those two RPCs are going to behave than it turned out in practice. Now it sure seems like they could be merged into a single `UpdateGateway` RPC.
   * Honestly, this occured to me only after I finished everything and started to create this PR. At this point I'd rather spend the time on delivering other features and IMHO we can postpone this refactor until we need to update a third field on the gateway (if ever).
* How do I test this by myself?
   * There's a `connect/db-port-wip` branch in the webapps repo. You can build Connect off that branch and after connecting to the db you should see a form field for changing the port number. Changing the value makes calls this new handler.

---

A small reminder that [lib/teleterm has a readme](https://github.com/gravitational/teleport/tree/ravicious/gateway-hash-map/lib/teleterm) in case this is your first foray into this directory.
